pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -605,25 +608,109 @@ private static File convertURIToSegmentLocation(String segmentLocation) {
+     * So when a new controller takes over as leader, or when validation manager runs, it needs to check the following:
+     *
+     * If it fails between step-1 and step-2:
+     * Check whether there are any segments in the PROPERTYSTORE with status DONE, but no new segment in status IN_PROGRESS,
+     * and hence the status of the segment in the IDEALSTATE is still CONSUMING
+     *
+     * If it fails between step-2 and-3:
+     * Check whether there are any segments in PROPERTYSTORE with status IN_PROGRESS, that are not accounted for in idealState.
+     * If so, it should create the new segments in idealState.
+     *
+     * If the controller fails after step-3, we are fine because the idealState has the new segments.
+     * If the controller fails before step-1, the server will see this as an upload failure, and will re-try.
+     */
+
+    final LLCSegmentName committingLLCSegmentName = new LLCSegmentName(committingSegmentNameStr);
+
+    // Step-1
+    boolean success = updateOldSegmentMetadataZNRecord(realtimeTableName, committingLLCSegmentName, nextOffset);
+    if (!success) {
+      return false;
+    }
+
+    final int partitionId = committingLLCSegmentName.getPartitionId();
+    final int newSeqNum = committingLLCSegmentName.getSequenceNumber() + 1;
+    final long now = System.currentTimeMillis();
+    LLCSegmentName newLLCSegmentName = new LLCSegmentName(committingLLCSegmentName.getTableName(), partitionId, newSeqNum, now);
+
+    final PartitionAssignment partitionAssignment = getStreamPartitionAssignment(realtimeTableName);
+    // If an LLC table is dropped (or cleaned up), we will get null here. In that case we should not be creating a new segment
+    if (partitionAssignment == null) {
+      LOGGER.warn("Kafka partition assignment not found for {}", realtimeTableName);
+      throw new RuntimeException("Kafka partition assignment not found. Not committing segment");
+    }
+
+    // Step-2
+    success = createNewSegmentMetadataZNRecord(realtimeTableName, newLLCSegmentName, nextOffset, partitionAssignment);
+    if (!success) {
+      return false;
+    }
+
+    // Step-3
+    // TODO Introduce a controller failure here for integration testing
+
+    // When multiple segments of the same table complete around the same time it is possible that
+    // the idealstate udpate fails due to contention. We serialize the updates to the idealstate
+    // to reduce this contention. We may still contend with RetentionManager, or other updates
+    // to idealstate from other controllers, but then we have the retry mechanism to get around that.
+    // hash code can be negative, so make sure we are getting a positive lock index
+
+    String newSegmentNameStr = newLLCSegmentName.getSegmentName();
+    List<String> newInstances = partitionAssignment.getInstancesListForPartition(String.valueOf(partitionId));
+
+    int lockIndex = (realtimeTableName.hashCode() & Integer.MAX_VALUE) % NUM_LOCKS;
```

Comment:
why do we need this lock?

Reply:
I haven't introduced the locking part. I have simply split the commitSegmentMetadata method into 3 parts 
1. update old segment metadata 
2. create new segment metadata 
3. update ideal state. 
I did that, so that it is easier to call just some of the parts from the code which tries to correct incomplete commits (completeCommittingSegment)

Reply:
there is a comment "when multiple segments of the same table..." . to justify the lock. It was added more than a year ago. without the lock, we were seeing each thread contending to update the idealstate, and backing off exponentially. Never saw the problem after the lock.

We can potentially batch the updates, but that is an optimization...